src/core/src/package_managers/YumPackageManager.py

```diff
@@ -282,7 +287,121 @@ def is_package_version_installed(self, package_name, package_version):
+        #Upgrading:
+        # polkit    x86_64 0.115-14.el8_6.1 rhel-8-for-x86_64-baseos-eus-rhui-rpms 154 k
+        # polkit-libs
+        #           x86_64 0.115-14.el8_6.1 rhel-8-for-x86_64-baseos-eus-rhui-rpms  77 k
+        #
+        #Transaction Summary
+        #================================================================================
+        #Upgrade  2 Packages
+        #
+        #Total download size: 231 k
+        #Operation aborted.
+
+        package_extensions = ["x86_64", "noarch", "i686"]
+        dependencies = []
+        lines = output.strip().split('\n')
+        upgrading_or_installing_flag = False
+
+        for line_index in range(0, len(lines)):
+            line = lines[line_index]
+            if not upgrading_or_installing_flag:
+                if not line.startswith("Upgrading:") and not line.startswith("Installing dependencies:"):
+                    self.composite_logger.log_debug(" - Inapplicable line: " + str(line))
+                    continue
+                else:
+                    upgrading_or_installing_flag = True
+                    self.composite_logger.log_debug(" - Packages list started: " + str(line))
+                    continue
+
+            if line.startswith("Transaction Summary"):
+                break
+
+            updates_line = re.split(r'\s+', line.strip())
+            dependent_package_name = ""
+
+            if len(updates_line) == 6 and updates_line[1] in package_extensions:
+                dependent_package_name = updates_line[0] + "." + updates_line[1]
+            elif len(updates_line) == 5 and line_index > 0 and updates_line[0] in package_extensions:
+                # Handling the scenario when package detail is spill over to 2 lines
+                prev_line = lines[line_index-1]
+                prev_updates_line = re.split(r'\s+', prev_line.strip())
+                if len(prev_updates_line) == 1:
+                    # Taking package name from previuos line and package extension from current line.
+                    dependent_package_name = prev_updates_line[0] + "." + updates_line[0]
+                else:
+                    # This is unexpected scenario.
+                    # If update details are divided in two lines then first line should contain package name and second should contain rest of the details.
+                    # Hence previous line should contain package name and current line should contain other 5 details.
+                    self.composite_logger.log_debug(" - Inapplicable line: " + str(line))
+                    continue
+            else:
+                self.composite_logger.log_debug(" - Inapplicable line: " + str(line))
+                continue
+
+            if len(dependent_package_name) != 0 and dependent_package_name not in packages:
+                self.composite_logger.log_debug(" - Dependency detected: " + dependent_package_name)
+                dependencies.append(dependent_package_name)
+
```

Comment:
I have some concerns with this implementation:

1.  The separation of logic for yum3 and yum4: While there are only 2 separate chunks right now, we cannot continue this line of implementation for each new version. There should be a generic way to handle this for all versions. 

2.  Adding to the first point: "All versions" includes implementing a solution that's easy to adapt to for future versions (i.e. any o/p format changes should be easy to incorporate in the existing logic)

3. Assumption on how the o/p text is structured: 

![image](https://github.com/Azure/LinuxPatchExtension/assets/56841542/c7929a92-9fb4-4e6b-a2f0-c241fad6e63a)
We are only considering these 2 samples with the code being designed with these fixed formats. Any change in these o/p formats would need a code change in future, which does not seem feasible.
for eg: 
            - what if the o/p is split in 3 lines? 
            - or the o/p is split in 2 lines but structured something like:
                        polkit-libs x86_64 0.115-14.el8_6.1
                                                                     rhel-8-for-x86_64-baseos-eus-rhui-rpms  77 k

Reply:
For point 1 and 2:
I tried searching other ways but could not find any. The only way I could figure out to fetch details is through parsing output of yum commands. I agree that it is not feasible to add implementation for every new YUM version with different output format but I think we don't have any other option currently.

For point 3:
I did test in multiple RHEL and Oracle VMs and could see only these formats. I have till now tested in 6-7 different VMs. Hence, I did implementation for the formats which I have observed. I agree that any change in these o/p formats would need a code change in future, but it will be very complicated to handle all the possible cases through parsing logic. The cases could be:
(a) o/p split in 3 or more lines.
(b) o/p split in 2 lines but structured differently and there can be many combinations of such cases.
It will be complicated to add parsing logic for all the cases as mentioned in (a) and (b)

The other output parsing logics in this repo are also based on observations and there can be issues if output is structured differently than expected.

Reply:
This still needs some more thought. I too will spend some time to think over it.

Reply:
Adding @kjohn-msft to the discussion...

Reply:
@kjohn-msft, @rane-rajasi  :: I wanted to share pros and cons of all the three approaches: approach in existing PR and the two approaches that @kjohn-msft has shared.

The two approaches that Koshy has shared addresses the point 3 but they do not address point 1 and 2 raised by Rajasi.

There are some Pros and Cons for these two approaches in compare to the existing logic in the PR.
Pros:
1. There is no need to find yum version in these two approaches. Finding yum version itself involves parsing output of a command and there could be different output format which are not observed during testing. So, it is good to avoid finding yum version.
2. There is single function for parsing output for both yum v3 and yum v4 output format.

Cons:
1. As there is single function for parsing both yum v3 and yum v4 output format, the function becomes little complex due to more conditions as it needs to handle both kind of output format.

Overall, I think that the approaches that Koshy has shared are better than existing approach in the PR. I will once analyze both approaches that Koshy has shared in terms of implementation as there are different output formats so need to take care of different formats and I will share my findings before raising new iteration with one of the two approaches.

Reply:
@GAURAVRAMRAKHYANI In my opinion, @kjohn-msft suggestions address all of my concerns. 

The main idea behind all 3 points I raised was that we should avoid introducing dependencies, have generic code  that covers all scenarios with as little or possibly no special handling

for eg: in the current PR, there is one block to address yum3 o/p, another to address yum4, later there might be another. This approach goes on to simply add lines of code, which is the code bloat part Koshy mentioned.

We should have a generic solution/function which is able to handle both types of o/p (yum3 and yum4) and should be scalable to address more such instances in future. (which was point 2 in my previous comment)

To summarize, what I'm leaning towards is a code that handles all types of o/p without being dependent on identifying yum versions or a only certain format of multiline o/p.

To move towards implementing these suggestions and removing dependencies (yum version and format independent multi-line), I believe needs a some code change in this PR. Let me know if you would be able to pick these up and when